File: 2563/two_pointers.py
from typing import *


class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        return self.countLessThan(nums, upper + 1) - self.countLessThan(nums, lower)

    def countLessThan(self, nums: List[int], upper: int) -> int:
        l, r = 0, len(nums) - 1
        count = 0
        while l < r:
            sum = nums[l] + nums[r]
            if sum < upper:
                count += r - l
                l += 1
            else:
                r -= 1
        return count


if __name__ == '__main__':
    sol = Solution()

    nums, lower, upper, expected = [0,1,7,4,4,5], 3, 6, 6
    assert sol.countFairPairs(nums, lower, upper) == expected

    nums, lower, upper, expected = [1,7,9,2,5], 11, 11, 1
    assert sol.countFairPairs(nums, lower, upper) == expected

File: 0016/two_pointers.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int threeSumClosest(vector<int>& nums, int target) {
    sort(nums.begin(), nums.end());
    size_t j, k;
    int closest = nums[0] + nums[1] + nums[2];
    for (size_t i = 0; i < nums.size() - 2; i++) {
      j = i + 1, k = nums.size() - 1;
      while (j < k) {
        int sum = nums[i] + nums[j] + nums[k];
        if (abs(sum - target) < abs(closest - target)) {
          closest = sum;
        }
        if (sum > target) { k--; }
        else { j++; }
       }
    }
    return closest;
  }
};

int main() {
  Solution sol;
  vector<int> nums;
  int target;
  int expected;

  nums = {-1, 2, 1, -4}, target = 1, expected = 2;
  EXPECT_EQ(sol.threeSumClosest(nums, target), expected);

  nums = {0, 0, 0}, target = 1, expected = 0;
  EXPECT_EQ(sol.threeSumClosest(nums, target), expected);

  nums = {1, 1, 1, 1}, target = -100, expected = 3;
  EXPECT_EQ(sol.threeSumClosest(nums, target), expected);

  exit(EXIT_SUCCESS);
}
